firebase_admin/db.py

```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
-            self._clients[base_url] = client
-        return self._clients[base_url]
+        emulator_host = os.environ.get(_EMULATOR_HOST_ENV_VAR)
+        if emulator_host:
+            if '//' in emulator_host:
+                raise ValueError(
+                    'Invalid {0}: "{1}". It must follow format "host:port".'.format(
+                        _EMULATOR_HOST_ENV_VAR, emulator_host))
+            self._emulator_host = emulator_host
+        else:
+            self._emulator_host = None
+
+    def get_client(self, db_url=None):
+        """Creates a client based on the db_url. Clients may be cached."""
+        if db_url is None:
+            db_url = self._db_url
+        base_url, params, use_fake_creds = \
+            _DatabaseService._parse_db_url(db_url, self._emulator_host)
+        if self._auth_override:
+            params['auth_variable_override'] = self._auth_override
+
+        client_cache_key = (base_url, json.dumps(params, sort_keys=True))
+        if client_cache_key not in self._clients:
+            if use_fake_creds:
+                credential = _EmulatorAdminCredentials()
+            else:
+                credential = self._credential.get_credential()
+            client = _Client(credential, base_url, self._timeout, params)
+            self._clients[client_cache_key] = client
+        return self._clients[client_cache_key]
 
     @classmethod
-    def _validate_url(cls, url):
-        """Parses and validates a given database URL."""
+    def _parse_db_url(cls, url, emulator_host=None):
+        """Parses a database URL into (base_url, query_params, use_fake_creds).
+
+        The input can be either a production URL (https://foo-bar.firebaseio.com/)
+        or an Emulator URL (http://localhost:8080/?ns=foo-bar). The resulting
+        base_url never includes query params. Any required query parameters will
+        be returned separately as a map (e.g. `{"ns": "foo-bar"}`).
+
+        If url is a production URL and emulator_host is specified, the result
+        base URL will use emulator_host, with a ns query parameter indicating
+        the namespace, parsed from the production URL. emulator_host is ignored
+        if url is already an emulator URL. In either case, use_fake_creds will
+        be set to True.
+        """
         if not url or not isinstance(url, six.string_types):
             raise ValueError(
                 'Invalid database URL: "{0}". Database URL must be a non-empty '
                 'URL string.'.format(url))
-        parsed = urllib.parse.urlparse(url)
-        if parsed.scheme != 'https':
-            raise ValueError(
-                'Invalid database URL: "{0}". Database URL must be an HTTPS URL.'.format(url))
-        elif not parsed.netloc.endswith('.firebaseio.com'):
+        parsed_url = urllib.parse.urlparse(url)
+        use_fake_creds = False
+        if parsed_url.netloc.endswith('.firebaseio.com'):
+            if parsed_url.scheme != 'https':
```

Comment:
In reply to this and the comment below to make `_parse_emulator_url` throw -- the two errors need `url` instead of `parsed_url`. Is it worth it to pass both `(parsed_url, url)` in just for the sake of throwing errors?